src/connector/src/sink/iceberg/mod.rs

```diff
@@ -1366,8 +1559,189 @@ impl SinkCommitCoordinator for IcebergSinkCommitter {
+
+        if self.is_exactly_once {
+            self.delete_row_by_sink_id_and_end_epoch(&self.db, self.sink_id, epoch)
+                .await?;
+            tracing::info!("Succeeded delete pre commit metadata in epoch {epoch}.");
+        }
+        Ok(())
+    }
+
+    async fn persist_pre_commit_metadata(
+        &self,
+        db: DatabaseConnection,
+        start_epoch: u64,
+        end_epoch: u64,
+        pre_commit_metadata: Vec<u8>,
+    ) -> Result<()> {
+        let m = exactly_once_iceberg_sink::ActiveModel {
+            sink_id: Set(self.sink_id),
+            end_epoch: Set(end_epoch),
+            start_epoch: Set(start_epoch),
+            metadata: Set(pre_commit_metadata),
+        };
+        exactly_once_iceberg_sink::Entity::insert(m)
+            .exec(&db)
+            .await?;
+        Ok(())
+    }
+
+    async fn delete_row_by_sink_id_and_end_epoch(
+        &self,
+        db: &DatabaseConnection,
+        sink_id: u32,
+        end_epoch: u64,
+    ) -> Result<()> {
+        let deleted_count = Entity::delete_many()
+            .filter(Column::SinkId.eq(sink_id))
+            .filter(Column::EndEpoch.eq(end_epoch))
+            .exec(db)
+            .await?
+            .rows_affected;
+
+        if deleted_count == 0 {
+            tracing::info!("No rows deleted.");
+        } else {
+            tracing::info!("Deleted {} row(s).", deleted_count);
+        }
+
         Ok(())
     }
+
+    async fn iceberg_sink_has_pre_commit_metadata(
+        &self,
+        db: &DatabaseConnection,
+        sink_id: u32,
+    ) -> Result<bool> {
+        let count = exactly_once_iceberg_sink::Entity::find()
+            .filter(exactly_once_iceberg_sink::Column::SinkId.eq(sink_id))
+            .count(db)
+            .await?;
+        Ok(count > 0)
+    }
+
+    async fn get_metadata_by_sink_id(
+        &self,
+        db: &DatabaseConnection,
+        sink_id: u32,
+    ) -> Result<BTreeMap<u64, Vec<u8>>> {
+        let models: Vec<Model> = Entity::find()
+            .filter(Column::SinkId.eq(sink_id))
+            .all(db)
+            .await?;
+
+        let mut result = BTreeMap::new();
+
+        for model in models {
+            result.insert(model.end_epoch, model.metadata);
+        }
+
+        Ok(result)
+    }
+
+    /// This is used to check whether all files in the current batch of pre-commits have been successfully submitted.
+    ///  It returns true if all filenames are present in the current snapshot, indicating successful committing.
+    async fn all_files_in_snapshot(
+        &self,
+        iceberg_config: &IcebergConfig,
+        write_results: Vec<IcebergCommitResult>,
+    ) -> Result<bool> {
+        let iceberg_common = iceberg_config.common.clone();
+        let table = iceberg_common
+            .load_table(&iceberg_config.java_catalog_props)
+            .await?;
+
+        if let Some(snapshot) = table.metadata().current_snapshot() {
```

Comment:
Currently, we only search files in the current snapshot, however, asynchronous compaction might create a new snapshot which might replace all files in the current snapshot. I think we need to find files in all snapshots (a little bit heavy).

Reply:
I think searching the snapshot files one by one is too heavy. It would be better if Iceberg-rs can let us to specify the snapshot id for a (Iceberg fast append) transaction. cc @ZENOTME  We can generate the snapshot id in advance and record it to the metastore. Later we can check whether it has been committed more easily.  https://github.com/risingwavelabs/iceberg-rust/blob/683fb89edeaf8d1baae69e1f376d68b92be1d496/crates/iceberg/src/transaction.rs#L139

Reply:
For this scene, instead of specifying the snapshot ID, can we just read back the snapshot ID? The snapshot ID will be generated before the transaction commit.